pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,472 @@
+			secretCoordinatesMatch := podSecret.UID == cachedSecret.UID &&
+				podSecret.Namespace == cachedSecret.Namespace &&
+				podSecret.Name == cachedSecret.Name
+
+			if hashesMatch {
+				if !secretCoordinatesMatch && len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit {
+					// While we're only matching at this point, we want to ensure this secret is considered valid in the future
+					// and so we make an additional write to the cache.
+					// writePulledRecord() is a noop in case the secret with the update hash already appears in the cache.
+					if err := f.writePulledRecord(image, imageRef, &kubeletconfiginternal.ImagePullCredentials{KubernetesSecrets: []kubeletconfiginternal.ImagePullSecret{podSecret}}); err != nil {
+						klog.ErrorS(err, "failed to write an image pulled record", "image", image, "imageRef", imageRef)
+					}
+				}
+				return false
+			}
+
+			if secretCoordinatesMatch {
+				if len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit { // we already know the hashes did not match, no need to check again
+					// While we're only matching at this point, we want to ensure the updated credentials are considered valid in the future
+					// and so we make an additional write to the cache.
+					// writePulledRecord() is a noop in case the hash got updated in the meantime.
+					if err := f.writePulledRecord(image, imageRef, &kubeletconfiginternal.ImagePullCredentials{KubernetesSecrets: []kubeletconfiginternal.ImagePullSecret{podSecret}}); err != nil {
+						klog.ErrorS(err, "failed to write an image pulled record", "image", image, "imageRef", imageRef)
+					}
+					return false
+				}
+			}
+		}
+	}
+
+	return true
+}
+
+func (f *PullManager) PruneUnknownRecords(imageList []string, until time.Time) {
+	f.pulledAccessors.GlobalLock()
+	defer f.pulledAccessors.GlobalUnlock()
+
+	pulledRecords, err := f.recordsAccessor.ListImagePulledRecords()
+	if err != nil {
+		klog.ErrorS(err, "failed to garbage collect ImagePulledRecords")
+		return
+	}
+
+	imagesInUse := sets.New(imageList...)
+	for _, ir := range pulledRecords {
+		if !ir.LastUpdatedTime.Time.Before(until) {
+			// the image record was only update after the GC started
+			continue
+		}
+
+		if imagesInUse.Has(ir.ImageRef) {
+			continue
+		}
+
+		if err := f.recordsAccessor.DeleteImagePulledRecord(ir.ImageRef); err != nil {
+			klog.ErrorS(err, "failed to remove an ImagePulledRecord", "imageRef", ir.ImageRef)
+		}
+	}
+
+}
+
+// initialize gathers all the images from pull intent records that exist
+// from the previous kubelet runs.
+// If the CRI reports any of the above images as already pulled, we turn the
+// pull intent into a pulled record and the original pull intent is deleted.
+//
+// This method is not thread-safe and it should only be called upon the creation
+// of the PullManager.
+func (f *PullManager) initialize(ctx context.Context) {
+	pullIntents, err := f.recordsAccessor.ListImagePullIntents()
```

Comment:
same comment here, if we get partial success here (pullIntents have some entries, and some errors were encountered decoding), I think we should log AND continue to clean up what we can